### DISCUSSION TITLE
Cant use pinia outside components.

### DISCUSSION THREAD

**Discussion by StefanGhiban** [vuejs/pinia#1157] [Help and Questions] [answered]
I have been trying to use pinia outside components without success. I have followed the guide from the docs, but the same result. If I follow the steps from the docs, I get into another error, which is: "Cannot access 'App' before initialization".
I want to use pinia outside a component in order to set authorization headers to an axios create instance.

**Comment by posva** [comment, marked as answer]
You have to make sure any useStore function is called after calling app.use(pinia). Maybe in your code the useStore function is being called before that.

**Reply by StefanGhiban** [reply]
Yep, that is the issue, I have found a workaround to make sure that does not happen. But having followed the guide from the docs, that does not seem to work for me, it doesnt really show how to use pinia outside a component properly. There is only one use case shown.

**Reply by Sovai** [reply]
Can you show me how you work around it? I cannot access pinia in router.js file even I called it after app.use(pinia)

**Reply by StefanGhiban** [reply]
Hi @Sovai ! In main.js/ts make sure to use pinia before router, this way you make sure that when the exported router from the router.js file is used, pinia will have already been instantiated. My workaround is not based on the router. I have tried to define some axios interceptors. The way I do it now is in a function that I export and call inside main.js. Doing things this way seems to work correctly.

**Reply by tomrob270** [reply]
@StefanGhiban Currently having this problem with axios interceptors, where defining the store within the axios interceptor method is throwing an error. 
Think it's because my pinia stores call the service file which in turn calls the pinia store so it trips up. Would you mind sharing how you fixed your problem?

**Reply by Sovai** [reply]
@tomrob270 Defined it inside the functIon instead of outside it will work
